overlord/devicestate/devicestate.go

```diff
@@ -784,15 +779,37 @@ func (r *remodeler) installedRevisionUpdateGoal(
 		return nil, fmt.Errorf("installed snap %q does not have the required revision in its sequence to be used for offline remodel: %s", sn.name, constraints.Revision)
 	}
 
-	// TODO:COMPS: snapstate currently reaches out to the store during a refresh
-	// if the snap has components already installed, regardless if the snap is
-	// already installed or not
-	if snapst.Sequence.HasComponents(index) {
-		return nil, errors.New("internal error: falling back to previous snap revision that has components installed is not supported during remodel")
+	ss := snapst.Sequence.Revisions[index]
+	comps := make([]snapstate.PathComponent, 0, len(ss.Components))
+	for _, c := range components {
+		cref := naming.NewComponentRef(snap.InstanceSnap(sn.name), c)
+		cs := ss.FindComponent(cref)
+		if cs == nil {
+			return nil, fmt.Errorf("cannot find required component in set of already installed components: %s", cref)
+		}
+
+		compConstraints := constraints.Component(cs.SideInfo.Component.ComponentName)
+		if !compConstraints.Revision.Unset() && compConstraints.Revision != cs.SideInfo.Revision {
+			return nil, fmt.Errorf("cannot fall back to component %q with revision %s, required revision is %s", cs.SideInfo.Component, cs.SideInfo.Revision, compConstraints.Revision)
+		}
+
+		cpi := snap.MinimalComponentContainerPlaceInfo(
+			cs.SideInfo.Component.ComponentName,
+			cs.SideInfo.Revision,
+			snapst.InstanceName(),
+		)
+
+		comps = append(comps, snapstate.PathComponent{
+			SideInfo: cs.SideInfo,
+			Path:     cpi.MountFile(),
+		})
 	}
 
-	return snapstateStoreUpdateGoal(snapstate.StoreUpdate{
+	return snapstatePathUpdateGoal(snapstate.PathSnap{
```

Comment:
will this actually move the old sequence point in front or will there be two sequence points?

Reply:
where to I see that is using a PathUpdateGoal?